### PR TITLE
storage: heuristic to detect topic recreation

### LINKED ISSUE #20062: sentry: panic: Attempted to downgrade Capability { time: Partitioned((Range(Elem(5), Top), MzOffset { offset: 0 ...
Sentry Issue: [DATABASE-BACKEND-2SJ](https://materializeinc.sentry.io/issues/4263712661/?referrer=github_integration)

```
panic: Attempted to downgrade Capability { time: Partitioned((Range(Elem(5), Top), MzOffset { offset: 0 })), internal: "..." } to Partitioned((Range(Elem(2), Top), MzOffset { offset: 0 })), which is not beyond the capability's time.
  File "panicking.rs", line 575, in rust_begin_unwind
  File "capability.rs", line 139, in timely::dataflow::operators::capability::Capability<T>::downgrade::downgrade_panic
  File "capability.rs", line 147, in timely::dataflow::operators::capability::Capability<T>::downgrade::{closure#0}
  File "capability.rs", line 146, in timely::dataflow::operators::capability::Capability<T>::downgrade
  File "kafka.rs", line 417, in mz_storage_client::types::sources::KafkaSourceConnection::render::
...
(37 additional frame(s) were not displayed)
```

### PR DESCRIPTION
**Pull request by petrosagg**
### Motivation

Kafka topics are identified by name but it's possible that a user recreates a topic with the same name but different configuration. Ideally we'd want to catch all of these cases and immediately error out the source, since the data is effectively gone. Unfortunately this is not possible without something like KIP-516[1] so we're left with heuristics.

This PR implements one such heuristic that checks if the reported number of partitions went down. If this happens we know that the topic got swapped out.

[1] https://cwiki.apache.org/confluence/display/KAFKA/KIP-516%3A+Topic+Identifiers

Fixes #20062

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by petrosagg** [comment]
@benesch I added a topic deletion command to testdrive but named it `kafka-delete-topic-flaky` so that it rings an alarm if anyone discovers it later.

**Comment by philip-stoev** [comment]
@petrosagg is it possible to also check if any of the per-partition offsets have gone backwards?

**Review by philip-stoev — APPROVED** [review]
Thank you for the test.

File: test/testdrive/kafka-recreate-topic.td [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+---
+1
+
+# Now recreate the topic with fewer partitions and observe the error
+
+$ kafka-delete-topic-flaky topic=topic0
+
+# Even though `kafka-delete-topic` ensures that the topic no longer exists in
+# the broker metadata there is still work to be done asychnronously before it's
+# truly gone that must complete before we attempt to recreate it. There is no
```
Comment (philip-stoev) [inline test/testdrive/kafka-recreate-topic.td:41]: You may wish to duplicate this comment in delete_topic.rs .

**Comment by philip-stoev** [comment]
Please update testdrive.md if you feel like it.

**Comment by petrosagg** [comment]
@philip-stoev updated testdrive.md and added high watermark regression detection (with a test)

**Comment by philip-stoev** [comment]
Thank you!

File: src/kafka-util/src/admin.rs [hunk 2]
```diff
@@ -140,3 +140,96 @@ pub enum CreateTopicError {
...
+    Retry::default()
+        .max_duration(Duration::from_secs(30))
+        .retry_async(|_| async {
+            let metadata = client
+                .inner()
+                // N.B. It is extremely important not to ask specifically
+                // about the topic here, even though the API supports it!
+                // Asking about the topic will create it automatically...
+                // with the wrong number of partitions. Yes, this is
+                // unbelievably horrible.
```
[the remark this answers is not in the document]
Comment (bkirwi) [reply src/kafka-util/src/admin.rs:211]: Have we reproduced this behaviour recently? In theory the client is only supposed to enable this behaviour when `allow.auto.create.topics` is set by the client: https://github.com/confluentinc/librdkafka/blob/master/CONFIGURATION.md
(I'm not sure what the default is for the admin client; we may need to set it explicitly to false.)

[the remark this answers is not in the document]
Comment (petrosagg) [reply src/kafka-util/src/admin.rs:211]: I haven't, this comment is copied as-is from the `create_topic` function of this crate which is probably a few years old

File: src/storage/src/source/source_reader_pipeline.rs [hunk 3]
```diff
@@ -1201,6 +1201,15 @@ async fn handle_message<K, V, T, D>(
...
 
+            match metric_updates.entry(partition) {
+                Entry::Occupied(mut entry) => {
+                    entry.insert((offset, ts, entry.get().2 + 1));
+                }
+                Entry::Vacant(entry) => {
+                    entry.insert((offset, ts, 1));
+                }
+            }
+
```
[the remark this answers is not in the document]
Comment (petrosagg) [reply src/storage/src/source/source_reader_pipeline.rs:1212]: Yeah but these metrics are very outdated, they assume that every source has partitions and offsets which is not true. We record message count in source statistics which sounds like the correct way to measure this. Eventually we should delete these metrics

**Pull request merged**